modules/utils/include/hephaestus/utils/bit_flag.h

```diff
@@ -0,0 +1,96 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <absl/synchronization/mutex.h>
+#include <fmt/core.h>
+#include <magic_enum_all.hpp>
+
+namespace heph::utils {
+
+template <typename EnumT>
+concept UnsignedEnum =
+    requires { std::is_enum_v<EnumT>&& std::is_unsigned_v<typename std::underlying_type<EnumT>::type>; };
+
+template <UnsignedEnum EnumT>
+constexpr auto checkEnumValuesArePowerOf2() -> bool {
+  constexpr auto POWER_OF_TWO = magic_enum::enum_for_each<EnumT>([](auto val) {
+    constexpr EnumT VALUE = val;
+    constexpr auto VALUE_T = static_cast<std::underlying_type<EnumT>::type>(VALUE);
+    return (VALUE_T & (VALUE_T - 1)) == 0;  // check if VALUE_T is power of 2
+  });
+
+  return std::all_of(POWER_OF_TWO.begin(), POWER_OF_TWO.end(), [](bool val) { return val; });
+}
+
+/// This class allows to use enum classes as bit flags.
+/// Enum classes need to satisfy three properties:
+/// - Underline type is unsigned (checked by template concept)
+/// - All values are power of 2 (checked by static assert)
+/// - No duplicated values (not checked, and impossible to until we get reflection)
+/// Usage:
+/// enum class Flag : uint8_t { A = 1u << 1u, B = 1u << 2u, C = 1u << 3u };
+/// BitFlag<Flag> flag{ Flag::A };
+/// flag.set(Flag::B).set(Flag::C);
+/// flag.has(Flag::B); // true
+/// flag.unset(Flag::A);
+/// Variable containing multiple flags can be created as constexpr:
+/// constexpr auto D = BitFlag<Flag>{ Flag::B }.set(Flag::C); // == Flag::B | Flag::C
+/// flag.hasAny(D); // true
+template <UnsignedEnum EnumT>
+class BitFlag {
+public:
+  // see https://dietertack.medium.com/using-bit-flags-in-c-d39ec6e30f08
+  using T = std::underlying_type<EnumT>::type;
+
+  constexpr explicit BitFlag(EnumT value) : value_{ value } {
+    static_assert(checkEnumValuesArePowerOf2<EnumT>(),
+                  "Enum is not valid for BitFlag, its values must be power of 2.");
+  }
+
+  /// Unset all flags and set the given flag.
+  constexpr auto reset(EnumT flag) -> BitFlag& {
```

Comment:
What is the use case of reset all, and then set one, in one function call? Alternative would be reset(), and then set(flag)

Reply:
I thought about that, but the flag needs to have at least one value, you cannot be empty, can you?

Reply:
Unless you have different information, I don't think any flag has to be set. If each flag signals an error, no flag means operational as desired. In that light, I would reset to 0. The current reset can nicely be achieved via reset().set(A) then

Reply:
actually, 0 is a valid value, that is correct, also if I set A and then unset A I would have no value. Updating